### PR TITLE
Bite 1632

### PR DESCRIPTION
**Pull request by bsomogyi**
- **What does the PR do?** *Adds MongoDB support in the form of a statefulset: https://agile-jira.pearson.com/browse/BITE-1632*
- **What automated testing exists for this change?** *Unit tests are part of the PR*
- **Is there documentation? **https://github.com/pearsontechnology/environment-operator/blob/bite-1632/docs/Mongo.md**  
- **Any dependencies?** *None*
- [x] **Has the changelog been updated?** 

For those on the Bitesize Pearson team wanting to test standing up a mongo statefulset you may follow the steps below:

1) Deploy a Bitesize VPC and PaaS from the Bitesize "dev' branch.
2) On the Master (once it is fully deployed), do the following:

```
aws s3 cp s3://bitesize-backups-dev/somogyi-temp/run-it.sh .
chmod 775 run-it.sh
./run-it.sh
```
3)Ensure the environment-operator pod is running and that the mongo stateful set is deployed. It will take a moment for the 3 mongo replicas to start. 
```
kubectl get pods --namespace=somogyi-app
```
4) Now configure the replicaset and add a user:

```
#Jump on a mongo pod
kubectl exec -it mongodb-0 --namespace=somogyi-app -c mongo bash

#Enter the mongo shell via localhost exception
mongo

#Initiate the replicaset
rs.initiate(
       {_id: "mongo", version: 1,members: [
       { _id: 0, host : "mongodb-0.mongodb.somogyi-app.svc.cluster.local:27017" },
       { _id: 1, host : "mongodb-1.mongodb.somogyi-app.svc.cluster.local:27017" },
       { _id: 2, host : "mongodb-2.mongodb.somogyi-app.svc.cluster.local:27017" },
       ],
       settings: {
          chainingAllowed : true, 
          heartbeatIntervalMillis : 2000,
          heartbeatTimeoutSecs: 10,
          electionTimeoutMillis : 10000,
          catchUpTimeoutMillis : 60000,
          getLastErrorModes : {},
          getLastErrorDefaults : {w: 'majority', wtimeout: 5000 }
       }}
);

#Ensure all members are in the replicaset
rs.status()

#Add a user to the admin database
db.getSiblingDB("admin").createUser({
      user : "admin",
      pwd  : "password",
      roles: [ { role: "root", db: "admin" } ]
 });

#Gain authorization as an administrator to add another user
db.getSiblingDB('admin').auth("admin", "password");

#swtich DBs
use nodeDB

#Add an appication user to a database
db.createUser({
      user : "user",
      pwd  : "abc123",
      roles : [ { "role" : "readWrite", "db" : "nodeDB" },
                { role: "dbAdmin", db: "nodeDB" } ]
  });

```
5) You can now deploy the sample app to see data may be added to the cluster
```
export endpoint=`kubectl get endpoints --namespace=somogyi-app | grep -i environment-operator | awk '{print $2}'`
export token=`cat ~/mongotest/token`

curl -k -XPOST -H "Authentication: Bearer ${token}" -H 'Content-Type: application/json' -d '{"application":"back", "name":"back", "version":"1.0.1-20171019164412"}' $endpoint/deploy
```
6) Verify inserts are working
```
kubectl get pods --namespace=somogyi-app
kubectl logs -f <back pod> --namespace=somogyi-app
```

### CODE REVIEW

File: docs/Mongo.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        port: 27017
+        volumes:
+          - name: mongo-persistent-storage
+            path: /data/db
+            modes: ReadWriteOnce
+            size: 10G
+        env:
+          - name: MONGO_PORT
+            value: 27017
+          - name: KUBE_NAMESPACE
```
Comment (geriBatai) [inline docs/Mongo.md]: why do we need this? Because we control the full pod construct here, we can just use [Kubernetes Downward API](https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/) environment variables to get current namespace into pod.

File: docs/Mongo.md [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            value: 'somogyi-app'
+          - name: MONGO_SIDECAR_POD_LABELS
+            value: "role=mongo,name=mongodb"
+          - name: KUBERNETES_MONGO_SERVICE_NAME
+            value: "mongodb"
+          - name: MONGO_DB_USERNAME
```
Comment (geriBatai) [inline docs/Mongo.md]: TBH, I would hide all these as well. We can hide MONGO_SIDECAR_POD_LABELS as well, and username, password and port should probably be presented as options fields in bitesize spec rather than allowing users access to these variables directly.

```
services:
- name: mongo
  database_type: mongodb
  version: 3.4
  options:
    # we could probably come up with references to kubernetes secrets here, instead of
    # plain-text passwords
    username: username
    password: password
```

I'd like to hide as much internals as possible from the end user

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -75,31 +75,46 @@ func (cluster *Cluster) ApplyEnvironment(currentEnvironment, newEnvironment *bit
...
 			}
 
-			deployment, err := mapper.Deployment()
-			if err != nil {
-				return err
-			}
-			if err = client.Deployment().Apply(deployment); err != nil {
-				log.Error(err)
-			}
+			if service.DatabaseType == "mongo" {
```
Comment (geriBatai) [inline pkg/cluster/cluster.go]: I think ifs in this place start requiring abstraction.

We have if service.Type, if service.DatabaseType and three different flows. Without much thinking, it looks like we should have an interface called ClusterService, with three implementations - ClusterMongoStatefulSet, ClusterDeployment, ClusterCRD (or clustertpr).

Then the flow would be something like:

```
clustersvc := service.GetClusterService() //returns appropriate ClusterService interface depending on service.Type or service.DatabaseType)
clustersvc.Apply() // performs mapper thingies
```

Not sure if it's too early to go this route, but it starts getting complex to follow.

Comment (bsomogyi) [reply pkg/cluster/cluster.go]: Not sure if the flow is over complicated and hard to follow at this point.  But since we do know there will be more types potentially coming down the pike that we may implement  I'll look at just abstracting it now to clean things up.

File: docs/Mongo.md (hunk 1 above)
Comment (bsomogyi) [reply docs/Mongo.md]: This is now OBE with ec74623

File: pkg/cluster/cluster.go (hunk 3 above)
Comment (bsomogyi) [reply pkg/cluster/cluster.go]: @geriBatai I have a few tickets coming  up around boiler plate cleanup/docs, which I'd like to do the refactor under (since its not necessary yet).  Let me know if this PR looks good as-is, knowing that I will capture that refactor in the near term.
